cli/js/testing.ts

```diff
@@ -120,70 +121,109 @@ export function test(
+  kind: "testStart";
+  test: TestDefinition;
 }
 
-enum TestStatus {
-  Passed = "passed",
-  Failed = "failed",
-  Ignored = "ignored"
-}
-
-interface TestResult {
+export interface TestEndMessage {
+  kind: "testEnd";
   name: string;
-  status: TestStatus;
+  status: "passed" | "failed" | "ignored";
   duration: number;
   error?: Error;
 }
 
-export enum TestEvent {
-  Start = "start",
-  TestStart = "testStart",
-  TestEnd = "testEnd",
-  End = "end"
+export interface RunTestsEndMessage {
+  kind: "runTestsEnd";
+  filtered: number;
+  ignored: number;
+  measured: number;
+  passed: number;
+  failed: number;
+  duration: number;
+  errors: Array<[string, Error]>;
 }
 
-interface TestEventStart {
-  kind: TestEvent.Start;
-  tests: number;
-}
+export type TestMessage =
+  | RunTestsStartMessage
+  | TestStartMessage
+  | TestEndMessage
+  | RunTestsEndMessage;
 
-interface TestEventTestStart {
-  kind: TestEvent.TestStart;
-  name: string;
-}
+const encoder = new TextEncoder();
+
+function log(msg: string, noNewLine = false): void {
+  if (!noNewLine) {
+    msg += "\n";
+  }
 
-interface TestEventTestEnd {
-  kind: TestEvent.TestEnd;
-  result: TestResult;
+  // Using `stdout` here because it doesn't force new lines
+  // compared to `console.log`; `core.print` on the other hand
+  // is line-buffered and doesn't output message without newline
+  stdout.writeSync(encoder.encode(msg));
 }
 
-interface TestEventEnd {
-  kind: TestEvent.End;
-  stats: TestStats;
-  duration: number;
-  results: TestResult[];
+function reportToConsole(message: TestMessage): void {
+  if (message.kind == "runTestsStart") {
+    log(`running ${message.tests.length} tests`);
+  } else if (message.kind == "testStart") {
+    const {
+      test: { name }
+    } = message;
+
+    log(`test ${name} ... `, true);
+    return;
+  } else if (message.kind == "testEnd") {
+    switch (message.status) {
+      case "passed":
+        log(`${GREEN_OK} ${formatDuration(message.duration)}`);
+        break;
+      case "failed":
+        log(`${RED_FAILED} ${formatDuration(message.duration)}`);
+        break;
+      case "ignored":
+        log(`${YELLOW_IGNORED} ${formatDuration(message.duration)}`);
+        break;
+    }
+  } else if (message.kind == "runTestsEnd") {
+    if (message.errors.length > 0) {
+      log(`\nfailures:\n`);
+
+      for (const [name, error] of message.errors) {
+        log(name);
+        log(stringifyArgs([error]));
+        log("");
+      }
+
+      log(`failures:\n`);
+
+      for (const [name] of message.errors) {
+        log(`\t${name}`);
+      }
+    }
+    log(
+      `\ntest result: ${message.failed ? RED_FAILED : GREEN_OK}. ` +
+        `${message.passed} passed; ${message.failed} failed; ` +
+        `${message.ignored} ignored; ${message.measured} measured; ` +
+        `${message.filtered} filtered out ` +
+        `${formatDuration(message.duration)}\n`
+    );
+  }
 }
 
+exposeForTest("reportToConsole", reportToConsole);
```

Comment:
IMO this is worse than exporting `ConsoleReporter` class explicitly, sure it's only used by `unit_test_runner` but `exposeForTest` exposes everything publicly, users can still access all of the properties (but they're obscured)

Reply:
This is more a comment on the `Deno[Deno.symbols.internal]` mechanism, which should be looked at independently. It won't allow an exploit or anything to let users use it if they really wanted to. What it does is declare contained functionalities as not formally considered part of the public API, i.e. not supported or stable.